File: code/graphs/prim.cpp
// Lazy Prim: grow one tree from s. A min-heap holds candidate crossing edges
// keyed by weight; pop the lightest, skip it if it now leads back into the tree,
// otherwise take it. adj[u] = u's weighted edges. Returns the MST total weight.
struct Edge { int to; long w; };

long prim(int V, vector<vector<Edge>>& adj, int s) {
    vector<bool> inTree(V, false);
    priority_queue<pair<long,int>, vector<pair<long,int>>,
                   greater<>> pq;                 // min-heap keyed by edge weight
    long total = 0; int picked = 0;
    inTree[s] = true;
    for (Edge e : adj[s]) pq.push({e.w, e.to});   // edges leaving the start
    while (picked < V - 1 && !pq.empty()) {
        auto [w, u] = pq.top(); pq.pop();
        if (inTree[u]) continue;                  // stale — both ends already in
        inTree[u] = true; total += w; picked++;   // take the crossing edge
        for (Edge e : adj[u])
            if (!inTree[e.to]) pq.push({e.w, e.to});
    }
    return total;
}

File: code/graphs/prim.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

// Lazy Prim from s. adj[u] holds (v, weight) pairs. Reverse turns the max-heap
// into a min-heap on weight; skip stale pops. Returns the MST total weight.
fn prim(n: usize, adj: &Vec<Vec<(usize, i64)>>, s: usize) -> i64 {
    let mut in_tree = vec![false; n];
    let mut pq = BinaryHeap::new();
    let (mut total, mut picked) = (0i64, 0usize);
    in_tree[s] = true;
    for &(v, w) in &adj[s] { pq.push(Reverse((w, v))); }
    while picked < n - 1 {
        let Some(Reverse((w, u))) = pq.pop() else { break };
        if in_tree[u] { continue; }           // stale — both ends already in
        in_tree[u] = true;                    // take the crossing edge
        total += w;
        picked += 1;
        for &(v, wt) in &adj[u] {
            if !in_tree[v] { pq.push(Reverse((wt, v))); }
        }
    }
    total
}
